src/db.c

```diff
@@ -1020,13 +1013,26 @@ void dictScanCallback(void *privdata, const dictEntry *de) {
 
 void hashtableScanCallback(void *privdata, void *entry) {
     scanData *data = (scanData *)privdata;
+    sds val = NULL;
+    sds key = NULL;
+
     robj *o = data->o;
     list *keys = data->keys;
     data->sampled++;
 
-    /* currently only implemented for SET scan */
-    serverAssert(o && o->type == OBJ_SET && o->encoding == OBJ_ENCODING_HASHTABLE);
-    sds key = (sds)entry; /* Specific for OBJ_SET */
+    /* This callback is only used for scanning elements within a key (hash
+     * fields, set elements, etc.) so o must be set here. */
+    serverAssert(o != NULL);
+
+    /* get key */
+    if (o->type == OBJ_SET && o->encoding == OBJ_ENCODING_HASHTABLE) {
+        key = (sds)entry;
+    } else if (o->type == OBJ_ZSET) {
```

Comment:
Ah, yes. Rather it should only check the type - we already know it uses hashtable encoding because we're in a hashtable callback, but we use the callback with a couple different types now.